data_structures/src/chain.rs

```diff
@@ -1520,14 +1521,38 @@ impl TransactionsPool {
                         let n_commits = dr_output.witnesses as usize;
 
                         if commits.len() >= n_commits {
-                            commits_vec.extend(
-                                commits
-                                    .drain()
-                                    .map(|(_h, c)| co_hash_index.remove(&c).unwrap())
-                                    .take(n_commits),
-                            );
-
-                            total_fee += dr_output.commit_fee * n_commits as u64;
+                            let filtered_commits: Vec<CommitTransaction> = commits
+                                .drain()
+                                .filter_map(|(_h, c)| {
+                                    co_hash_index.remove(&c).and_then(|commit_tx| {
+                                        let mut valid = true;
+                                        let mut hs_current_inputs = HashSet::new();
+                                        for input in commit_tx.body.collateral.iter() {
+                                            // This validations is to avoid commits with the same Input
+                                            if hs_inputs.contains(input)
+                                                || hs_current_inputs.contains(input)
+                                            {
+                                                valid = false;
```

Comment:
We can stop processing the inputs after the first invalid one, right?

```suggestion
                                                valid = false;
                                                break;
```